### ISSUE TITLE
Twisted's coverage reports should include the scripts that are run in subprocesses

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8266]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8266|
|Type|enhancement|
|Created|2016-04-01 05:48:06Z|
|Branch|https://github.com/twisted/twisted/tree/cov-subprocess-8266-3|

With the Python 3 on Windows port, some of these will need to be changed. We should have these be covered by coverage, so we're sure we're not breaking things.

<details><summary>Searchable metadata</summary>

```
trac-id__8266 8266
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__branches_cov_subprocess_8266_3 branches/cov-subprocess-8266-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1459489686793615 1459489686793615
changetime__1462722834088906 1462722834088906
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [47262]) Branching to cov-subprocess-8266.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [47289]) Branching to cov-subprocess-8266-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

This is good, I think.

The tox env now passes the correct environment variables, and copies out the coverage files from the tox env once it's run. The user now has to run coverage combine/report, but we shouldn't document our coverage testing process until things like https://bitbucket.org/hpk42/tox/issues/275/allow-running-commands-before-and-after are fixed and it makes sense.

See the 12.04 builder https://buildbot.twistedmatrix.com/builders/ubuntu12.04-py2.7-coverage/builds/22, which runs the tox env. After this lands, all builders will be migrated to it.

Reasons for some of the changes:

1. To make coverage collection + reporting work, we can't just execute the data file. We have to do it with ```-m``` so it recognises that it is part of the Twisted package. This requires passing through the environment properly.

2. The ```-u``` makes Python use unbuffered I/O, which makes some tests less flaky.

3. The ```.pth``` hack is from http://nedbatchelder.com/blog/201001/running_code_at_python_startup.html .

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Can we have a comment on admin/_copy.py describing its role

same for admin/zz_coverage.pth

Rather than having a comment in the trac ticket about `the .pth hack` I would prefer to have it as a comment inside the file.

-------

I am not very happy with the way we use tox. it looks like we are using it both as a test runner and as a build tool.

but I am happy to see that we have removed some test related logic from buildbot and have it inside the code branch.

------

Can we have a simple `./admin/ci.py` script which is called by buildbot using various arguments.

so instead of `3 clearing virtualenv + 4 installing tox + 5 clearing coverage`, build will just call `./admin/ci.py before` ...

similar instead of `7 run coverage combine + 8 run coverage html + 10 run coverage xml + 11 upload to codecov` maybe we can have `./admin/ci.py after`

the upload can remain as a buildbot step

------

What is `properEnv` and why do we want to use a non-properEnv ? :)

-------

I would say that this can be merged and I am happy to see more test running logic contained in the branch.

Leaving this for review so that another dev can check this.

For me it looks good to merge... with the above minor comments.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @hawkowl|
|-|-|

Abandon your fear, adi!

I am removing this from review because it has feedback, and we should get out of the habit of passing the buck to other developers.  I haven't looked at it, I assume that adi's feedback is sensible.

Hawkowl, please address the points raised above and resubmit.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl removed owner|
|-|-|

Hi, thanks for the review.

1. Comment added about the ```pth``` file in it, and in the ```_copy.py```.

2. I think we're using it reasonably; it's an automation tool, not just a dedicated test runner. Build tasks are part of its documented use.

3. I don't know about a ```ci.py``` - the good thing about them being encoded in the buildbot (and later, the travis config) is that then the individual steps have logs and it's not just a massive dump to the stdio.

4. Added a comment about ```properEnv``` to the relevant files.

As per Glyph, putting back up for review. Builders spun.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks!

2. I am not happy with tox, since to automate things you need hacks like `admin/_copy.py` and I think that the .ini format is abused here.

Since we still delegate to external python scripts and hacks, why not replace

```
    coverage: python {toxinidir}/admin/_copy.py {toxinidir}/admin/zz_coverage.pth {envsitepackagesdir}/zz_coverage.pth
    coverage: coverage erase
    coverage: coverage run -p --rcfile={toxinidir}/.coveragerc {envbindir}/trial --reactor={env:TWISTED_REACTOR:default} {posargs:twisted}
    coverage: python {toxinidir}/admin/_copy.py "_trial_temp/.coverage*" {toxinidir}
    coverage: python {toxinidir}/admin/_copy.py ".coverage*" {toxinidir}
```

with just

```
    coverage: ./admin/coverage --test={posargs:twisted}
```

in this way tox.ini is our CI entry point with a clean delegation

-----------

3. I think that we can work around a convention for buildbot to break each step into separate logs

The ugly part about buildbot steps is that if you want to add or modify a step you need to go the twisted-admin route ... and manual testing if that step is much complicated 

With a simple script, you can just set up the CI env on your local system and try to run the command in a similar way as buildbot.

-----

4. Maybe `properEnv` can be renamed to `bytesOnlyEnv` or something more descriptive.

-------

Good to merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [47360]) Branching to cov-subprocess-8266-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47362]) Merge cov-subprocess-8266-3: Twisted's coverage reports should include the scripts that are run in subprocesses

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8266](https://github.com/twisted/twisted/issues/8266)
